src/commands/api/findTreeItem.ts

```diff
@@ -3,57 +3,87 @@
-    const databaseHostPort = await getHostPortFromConnectionString(connectionString);
-    const accountHostPort = await getHostPortFromConnectionString(account.connectionString);
-    return (databaseHostPort.host === accountHostPort.host && databaseHostPort.port === accountHostPort.port);
-}
-
-export async function findTreeItem(query: TreeItemQuery, attachedAccountsNode: AttachedAccountsTreeItem): Promise<DatabaseTreeItem | undefined> {
+export async function findTreeItem(query: TreeItemQuery): Promise<DatabaseTreeItem | undefined> {
     const connectionString = query.connectionString;
+    if (!/^mongodb[^:]*:\/\//i.test(connectionString)) {
+        throw new Error('Method not implemented for non-mongo connection string.');
+    }
 
-    if (/^mongodb[^:]*:\/\//i.test(connectionString)) {
-        const hostport = await getHostPortFromConnectionString(connectionString);
-        const accounts = await attachedAccountsNode.getCachedChildren();
-        for (const account of accounts) {
-            if ((account instanceof MongoAccountTreeItem) && (await isParentAccount(connectionString, account))) {
-                const databases = await account.getCachedChildren();
-                for (const database of databases) {
-                    if ((database instanceof MongoDatabaseTreeItem) && getDatabaseNameFromConnectionString(connectionString) === getDatabaseNameFromConnectionString(database.connectionString)) {
-                        return {
-                            connectionString: connectionString,
-                            databaseName: database.databaseName,
-                            hostName: hostport.host,
-                            port: hostport.port,
-                            reveal: async () => await ext.treeView.reveal(database)
-                        };
-                    }
+    const parsedCS = await parseMongoConnectionString(connectionString);
+
+    if (!parsedCS.databaseName) {
+        throw new Error('Method not implemented for account-level connection string.');
+    }
+
+    const maxTime = Date.now() + 10 * 1000; // Give up searching subscriptions after 10 seconds and just attach the account
+
+    // 1. Get result from cache if possible
+    let result: DatabaseTreeItem | undefined = tryGetTreeItemFromCache(parsedCS);
+
+    // 2. Search attached accounts (do this before subscriptions because it's faster)
+    if (!result) {
+        const attachedDbAccounts = await ext.attachedAccountsNode.getCachedChildren();
+        result = await searchDbAccounts(attachedDbAccounts, parsedCS, maxTime);
+    }
+
+    // 3. Search subscriptions
+    if (!result) {
+        const rootNodes = await ext.tree.getChildren();
+        for (const rootNode of rootNodes) {
+            if (Date.now() > maxTime) {
+                break;
+            }
+
+            if (rootNode instanceof CosmosDBAccountProvider) {
+                const dbAccounts = await rootNode.getCachedChildren();
```

Comment:
getCachedChildren means we won't ask Azure for more nodes?

Reply:
Correct. It will ask for the first batch of nodes if they haven't been retrieved yet, but that's it.